internal/executor/service/pod_issue_handler.go

```diff
@@ -0,0 +1,348 @@
+
+				issue := &podIssue{
+					OriginalPodState: pod.DeepCopy(),
+					JobId:            util.ExtractJobId(pod),
+					RunId:            util.ExtractJobRunId(pod),
+					Message:          message,
+					Retryable:        retryable,
+					Type:             podIssueType,
+				}
+				p.registerIssue(issue)
+				break
+			}
+		}
+	}
+}
+
+func (p *PodIssueService) handleKnownPodIssues(ctx context.Context, allManagedPods []*v1.Pod) {
+	issues := createIssues(allManagedPods, p.knownPodIssues)
+	// Make issues from pods + issues
+	util.ProcessItemsWithThreadPool(ctx, 20, issues, p.handlePodIssue)
+}
+
+func createIssues(managedPods []*v1.Pod, podIssues map[string]*podIssue) []*issue {
+	podsByRunId := make(map[string]*v1.Pod, len(managedPods))
+
+	for _, pod := range managedPods {
+		runId := util.ExtractJobRunId(pod)
+		if runId != "" {
+			podsByRunId[runId] = pod
+		}
+	}
+
+	result := make([]*issue, 0, len(podIssues))
+
+	for _, podIssue := range podIssues {
+		relatedPod := podsByRunId[podIssue.RunId]
+		result = append(result, &issue{CurrentPodState: relatedPod, Issue: podIssue})
+	}
+
+	return result
+}
+
+func (p *PodIssueService) handlePodIssue(issue *issue) {
+	// Skip jobs with no issues
+	if issue == nil {
+		return
+	}
+
+	hasSelfResolved := hasPodIssueSelfResolved(issue)
+	if hasSelfResolved {
+		p.markIssuesResolved(issue.Issue)
+		return
+	}
+
+	if issue.Issue.Retryable {
+		p.handleRetryableJobIssue(issue)
+	} else {
+		p.handleNonRetryableJobIssue(issue)
+	}
+}
+
+// For non-retryable issues we must:
+//   - Report JobUnableToScheduleEvent if the issue is a startup issue
+//   - Report JobFailedEvent
+//
+// Once that is done we are free to cleanup the pod
+func (p *PodIssueService) handleNonRetryableJobIssue(issue *issue) {
+	if !issue.Issue.Reported {
+		message := issue.Issue.Message
+
+		events := make([]reporter.EventMessage, 0, 2)
+		if issue.Issue.Type == StuckStartingUp || issue.Issue.Type == UnableToSchedule {
+			unableToScheduleEvent := reporter.CreateJobUnableToScheduleEvent(issue.Issue.OriginalPodState, message, p.clusterContext.GetClusterId())
+			events = append(events, reporter.EventMessage{Event: unableToScheduleEvent, JobRunId: issue.Issue.RunId})
+		}
+		failedEvent := reporter.CreateSimpleJobFailedEvent(issue.Issue.OriginalPodState, message, p.clusterContext.GetClusterId(), issue.Issue.Cause)
+		events = append(events, reporter.EventMessage{Event: failedEvent, JobRunId: issue.Issue.RunId})
+
+		err := p.eventReporter.Report(events)
+		if err != nil {
+			log.Errorf("Failed to report failed event for job %s because %s", issue.Issue.JobId, err)
+			return
+		}
+
+		p.markIssueReported(issue.Issue)
+	}
+
+	if issue.CurrentPodState != nil {
+		p.clusterContext.DeletePods([]*v1.Pod{issue.CurrentPodState})
+	} else {
+		p.markIssuesResolved(issue.Issue)
+	}
+}
+
+// For retryable issues we must:
+//   - Report JobUnableToScheduleEvent
```

Comment:
I'm not sure the scheduler cares about `unableToScheduleEvent`.  What's this used for?

Reply:
It is informational

Really it should just be "StartUpIssueEvent"

We could remove it - and we probably should if we had better state management. 

It is sometimes helpful when debugging to see that the issue was detected and attempted to be handled
 - Especially as part of handling the issue is deleting the pod - which could mean the executor loses state during this process. (i.e executor deletes it and then shuts down before it can return lease)